spec/src/antlr/JQL-full.g4

```diff
@@ -0,0 +1,769 @@
+    '(' string_expression 'AS' ('INTEGER' | 'LONG' | 'FLOAT' | 'DOUBLE') ')'
+    ;
+
+functions_returning_numerics
+    : 'LENGTH' '(' string_expression ')'
+    | 'LOCATE' '(' string_expression ',' string_expression (',' arithmetic_expression)? ')'
+    | 'ABS' '(' arithmetic_expression ')'
+    | 'CEILING' '(' arithmetic_expression ')'
+    | 'EXP' '(' arithmetic_expression ')'
+    | 'FLOOR' '(' arithmetic_expression ')'
+    | 'LN' '(' arithmetic_expression ')'
+    | 'SIGN' '(' arithmetic_expression ')'
+    | 'SQRT' '(' arithmetic_expression ')'
+    | 'MOD' '(' arithmetic_expression',' arithmetic_expression ')'
+    | 'POWER' '(' arithmetic_expression',' arithmetic_expression ')'
+    | 'ROUND' '(' arithmetic_expression',' arithmetic_expression ')'
+    | 'SIZE' '(' collection_valued_path_expression ')'
+    | 'INDEX' '(' identification_variable ')'
+    | extract_datetime_field
+    ;
+
+functions_returning_datetime
+    : 'CURRENT_DATE'
+    | 'CURRENT_TIME'
+    | 'CURRENT_TIMESTAMP'
+    | 'LOCAL' 'DATE'
+    | 'LOCAL' 'TIME'
+    | 'LOCAL' 'DATETIME'
+    | extract_datetime_part
+    ;
+
+string_cast_function
+    : 'CAST' '(' scalar_expression 'AS' 'STRING' ')'
+    ;
+
+functions_returning_strings
+    : 'CONCAT' '(' string_expression ',' string_expression (',' string_expression)* ')'
+    | 'SUBSTRING' '(' string_expression ',' arithmetic_expression (',' arithmetic_expression)? ')'
+    | 'TRIM' '(' (trim_specification? trim_character? 'FROM')? string_expression ')'
+    | 'LOWER' '(' string_expression ')'
+    | 'UPPER' '(' string_expression ')'
+    ;
+
+trim_specification
+    : 'LEADING'
+    | 'TRAILING'
+    | 'BOTH'
+    ;
+
+function_invocation
+    : 'FUNCTION'
+      '(' function_name (',' function_arg)* ')'
+    ;
+
+extract_datetime_field
+    : 'EXTRACT'
+      '(' datetime_field 'FROM' datetime_expression ')'
+    ;
+
+datetime_field
+    : identification_variable
+    ;
+
+extract_datetime_part
+    : 'EXTRACT'
+      '(' datetime_part 'FROM' datetime_expression ')'
+    ;
+
+datetime_part
+    : identification_variable
+    ;
+
+function_arg
+    : literal
+    | state_valued_path_expression
+    | input_parameter
+    | scalar_expression
+    ;
+
+entity_id_or_version_function
+    : id_function
+    | version_function
+    ;
+
+id_function
+    : 'ID'
+      '(' (general_identification_variable | single_valued_object_path_expression) ')'
+    ;
+
+version_function
+    : 'VERSION'
+      '(' (general_identification_variable | single_valued_object_path_expression) ')'
+    ;
+
+case_expression
+    : general_case_expression
+    | simple_case_expression
+    | coalesce_expression
+    | nullif_expression
+    ;
+
+general_case_expression
+    : 'CASE' when_clause+
+      'ELSE' scalar_expression
+      'END'
+    ;
+
+when_clause
+    : 'WHEN' conditional_expression
+      'THEN' scalar_expression
+    ;
+
+simple_case_expression
+    : 'CASE' case_operand simple_when_clause+
+      'ELSE' scalar_expression
```

Comment:
I believe the `ELSE` should be optional, but I am not sure.

Reply:
Hrrm, good catch. It's certainly optional in HQL, so I guess you're right. But it's marked required in JPQL.

So best if we open an issue for that.

Reply:
Done: https://github.com/jakartaee/persistence/issues/742